sdk/python/kfp/v2/google/aiplatform.py

```diff
@@ -0,0 +1,464 @@
+      of this field will be propagated to all worker pools if not specified
+      otherwise in additional_job_spec.
+    additional_job_spec: Full-fledged custom job API spec. The value specified
+      in this field will override the defaults provided through other function
+      parameters.
+
+      For details please see:
+      https://cloud.google.com/ai-platform-unified/docs/reference/rest/v1beta1/CustomJobSpec
+
+  Returns:
+    A KFP ContainerOp object represents the launcher container job, from which
+    the user training program will be submitted to AI Platform (Unified) Custom
+    Job service.
+
+  Raises:
+    KeyError on name collision between parameter and artifact I/O declaration.
+    ValueError when:
+      1. neither or both image_uri and executor_image_uri are provided; or
+      2. no valid package_uris and python_module is provided for custom Python
+         training.
+  """
+  # Check the sanity of the provided parameters.
+  input_artifacts = input_artifacts or {}
+  input_parameters = input_parameters or {}
+  output_artifacts = output_artifacts or {}
+  output_parameters = output_parameters or {}
+  if bool(set(input_artifacts.keys()) & set(input_parameters.keys())):
+    raise KeyError('Input key conflict between input parameters and artifacts.')
+  if bool(set(output_artifacts.keys()) & set(output_parameters.keys())):
+    raise KeyError('Output key conflict between output parameters and '
+                   'artifacts.')
+
+  if not additional_job_spec and  bool(image_uri) == bool(executor_image_uri):
+    raise ValueError('The user program needs to be either a custom container '
+                     'training job, or a custom Python training job')
+
+  # For Python custom training job, package URIs and modules are also required.
+  if executor_image_uri:
+    if not package_uris or not python_module or len(package_uris) > 100:
+      raise ValueError('For custom Python training, package_uris with length < '
+                       '100 and python_module are expected.')
+
+  # Check and scaffold the parameters to form the custom job request spec.
+  custom_job_spec = additional_job_spec or {}
+  if not custom_job_spec.get('workerPoolSpecs'):
+    # Single node training, deriving job spec from top-level parameters.
+    if image_uri:
+      # Single node custom container training
+      worker_pool_spec = {
+          "machineSpec": {
+              "machineType": machine_type or "n1-standard-4"
+          },
+          "replicaCount": "1",
+          "containerSpec": {
+              "imageUri": image_uri,
+          }
+      }
+      if commands:
+        worker_pool_spec['containerSpec']['command'] = commands
+      if args:
+        worker_pool_spec['containerSpec']['args'] = args
+      custom_job_spec['workerPoolSpecs'] = [worker_pool_spec]
+    if executor_image_uri:
+      worker_pool_spec = {
+          "machineSpec": {
+              "machineType": machine_type or "n1-standard-4"
```

Comment:
nit: make a constant for the default machine type?

Reply:
Done. Actually let's change default to n1-standard-2 to be more cost-effective